### PR TITLE
Add: Undo 'git tag'

### PR DESCRIPTION
**Pull request by kopalchakravarty**
@Bhupesh-V 

Adding functionality to undo git tag command 

**Demo:** ugit - undo git tag 

https://user-images.githubusercontent.com/31576619/183704532-a77cba9d-a061-4e17-a598-e17951d0c6e4.mp4

### CODE REVIEW

File: ugit [hunk 1]
```diff
@@ -266,6 +267,29 @@ undo_all() {
...
+undo_git_tag() {
+    # Fetch all tags from remote
+    if git fetch --all --tags > /dev/null 2>&1; then
+        printf "\n Fetching tags from remote"
+    else
+        perror "Unable to fetch tags from remote.Please check repository access."
+        exit
+    fi
+    tag=$(git tag | fzf --ansi --height 80% \
+        --reverse --multi --header="Choose tag to undo" \
```
Comment (Bhupesh-V) [inline ugit:279]:
```suggestion
        --reverse --multi --header="Choose a tag to remove" \
```

File: ugit [hunk 2]
```diff
@@ -266,6 +267,29 @@ undo_all() {
     fi
 }
 
+undo_git_tag() {
+    # Fetch all tags from remote
+    if git fetch --all --tags > /dev/null 2>&1; then
+        printf "\n Fetching tags from remote"
```
Comment (Bhupesh-V) [inline ugit:273]: Remove un-necessary whitespace

```suggestion
        printf "\nFetching tags from remote"
```

File: ugit [hunk 3]
```diff
@@ -266,6 +267,29 @@ undo_all() {
...
+        --preview "echo {} | cut -d' ' -f1 | xargs -I{} git show --color --pretty=format:%b {}" \
+        --bind 'j:down,k:up,ctrl-j:preview-down,ctrl-k:preview-up,ctrl-space:toggle-preview' --preview-window right:60% )
+
+    # Exit if no tag is selected
+    [ -z "$tag" ] && exit
+
+    # Delete git tag
+    printf "%s\nDeleting tag: $tag"
+    git push origin --delete "$tag" > /dev/null 2>&1
+    git tag -d "$tag" > /dev/null 2>&1
```
Comment (Bhupesh-V) [inline ugit:289]: Add both these statements in `if` safe check. 
1. Remove it locally, if it succeds only then remove it from Remote

File: ugit [hunk 4]
```diff
@@ -266,6 +267,29 @@ undo_all() {
...
+    echo && printf "%s\nTag $tag deleted 👍️\n"
```
Comment (Bhupesh-V) [inline ugit:290]: What's the use of this extra `echo` here?

File: ugit [hunk 5]
```diff
@@ -266,6 +267,29 @@ undo_all() {
...
+    tag=$(git tag | fzf --ansi --height 80% \
+        --reverse --multi --header="Choose tag to undo" \
+        --preview "echo {} | cut -d' ' -f1 | xargs -I{} git show --color --pretty=format:%b {}" \
+        --bind 'j:down,k:up,ctrl-j:preview-down,ctrl-k:preview-up,ctrl-space:toggle-preview' --preview-window right:60% )
+
+    # Exit if no tag is selected
+    [ -z "$tag" ] && exit
+
+    # Delete git tag
+    printf "%s\nDeleting tag: $tag"
```
Comment (Bhupesh-V) [inline ugit:287]: Let's add a warning prompt here. Something like this

`Undoing a git tag, removes the git tag from both your local repo as well as the remote origin. Continue [N/y]: `

now if the user enters `y` or `Y`, go on with the rest of the flow

**Review by Bhupesh-V — CHANGES_REQUESTED** [review]

File: ugit (hunk 4 above)
Comment (kopalchakravarty) [reply ugit:290]: Running echo just to insert a blank line. Should I instead add another new line in the printf command itself?

Comment (Bhupesh-V) [reply ugit]: Yes using 1 form for printing stuff is good to maintain consistency, We either use `echo` everywhere or `printf`

File: ugit [hunk 6]
```diff
@@ -266,6 +267,39 @@ undo_all() {
...
+    # Delete git tag
+    if [[ "$USER_INPUT" == Y || "$USER_INPUT" == y ]] && printf "%s\nDeleting tag: $tag"; then
+        if  git tag -d "$tag" > /dev/null 2>&1; then
+            git push origin --delete "$tag" > /dev/null 2>&1
+            printf "%s\n\nTag $tag deleted 👍️\n"
+        else
+            perror "Failed to delete tag."
+        fi
+    else
+        perror "User cancelled operation. Exiting ..."
```
[the remark this answers is not in the document]
Comment (Bhupesh-V) [reply ugit:298]: Looks perfect 🔥 , would you mind adding a short video demo to the PR itself, would be good for futute in case we need to debug something

**Comment by Bhupesh-V** [comment]
@kopalchakravarty One minor thing I forgot to mentions. can you have a look at the failed ShellCheck CI test and make changes accordingly?
https://github.com/Bhupesh-V/ugit/runs/7746413382?check_suite_focus=true

Its a issue about using the `read` command.

**Comment by kopalchakravarty** [comment]
> @kopalchakravarty One minor thing I forgot to mentions. can you have a look at the failed ShellCheck CI test and make changes accordingly? https://github.com/Bhupesh-V/ugit/runs/7746413382?check_suite_focus=true
> 
> Its a issue about using the `read` command.

Fixed it

File: ugit (hunk 6 above)
[the remark this answers is not in the document]
Comment (kopalchakravarty) [reply ugit]: Added the video

**Comment by Bhupesh-V** [comment]
Looks prefect. For upcoming PRs make sure to write good commit messages. For e.g instead of using `ugit: undo git tag` repeatedly use something like `remove uncessary error report if user cancels the operation`, `fix shellcheck read warning` etc.

Other than great work 🔥

**Comment by Bhupesh-V** [comment]
@all-contributors please add @kopalchakravarty  for code

**Pull request merged**